### PR TITLE
chore: add workflow for creating and storing SBOM

### PR DESCRIPTION
**Pull request by siberov**
This new workflow creates a software bill of materials (SBOM) formatted as an in-toto attestation. This covers the content of the source tree, as well as the compiled files. The SBOM is pushed to a remote repository under the chains-project organisation for long-term storage.

**NOTE:** To work, this requires that a secret called `SBOM_DEPLOY_SSH` is added to the Spoon repo. It must contain the private part of an SSH key, where the public key is added as a deploy key to the [chains-project/sbom-files](https://github.com/chains-project/sbom-files) repository.

This should not be merged before the note above has been addressed.

### CODE REVIEW

**Comment by monperrus** [comment]
FYI, this work is part of the Chains research project, where we push the state of the art of software supply chain engineering, see https://chains.proj.kth.se/

File: .github/workflows/sbom.yml
Comment (algomaster99) [inline .github/workflows/sbom.yml:69]:
```suggestion
          mkdir -p sbom-files/spoon/slsa
```

You can nest it another level so that we can store SBOM/provenance files generated for Spoon by other tools as well. Proposed structure:
```
sbom-files/
    <project>
        <sbom-generator-tool>
            *sbom.json
            *.provenance
```

File: .github/workflows/sbom.yml
Comment (algomaster99) [inline .github/workflows/sbom.yml:70]:
```suggestion
          cp build.provenance sbom-files/spoon/slsa/$(date +%s)-${{ github.sha }}.provenance
```

**Review by algomaster99 — CHANGES_REQUESTED** [review]

File: .github/workflows/sbom.yml
Comment (siberov) [reply .github/workflows/sbom.yml:70]: Good suggestion. Regarding this though, if we want to have a common naming convention: I'm naming the files <timestamp in seconds>-<commit sha>.provenance. The main idea being that it would be easy to find the relevant file, and having them sorted by time by default. Is this a good idea, or do we have another naming convention?

Comment (siberov) [reply .github/workflows/sbom.yml:70]: I have not seen it recommended anywhere, however, it is the name given in the example of the bottom of the README here: https://github.com/slsa-framework/github-actions-demo. I copied it since it made enough sense to me. :)

Comment (siberov) [reply .github/workflows/sbom.yml:70]: Sounds good to me, it is probably also clearer for humans when it is explicitly said to be a json file.

Comment (algomaster99) [reply .github/workflows/sbom.yml:70]: > clearer for humans when it is explicitly said to be a JSON file.

Fully agree.

Comment (siberov) [reply .github/workflows/sbom.yml:70]: > mistake `.provenance` file to be an executable file

BTW, I think that that is caused by the file being too large, not the extension. But I still think that `.json` is a more appropriate extension.

Comment (algomaster99) [reply .github/workflows/sbom.yml]: > I think that that is caused by the file being too large, not the extension

You may be right. Let me try it. Thanks for bringing this to my notice. :)

**Comment by MartinWitt** [comment]
I know this is scientific important, but I remember last time we had source graph builds in our CI which often failed and were broken for a longer time. How stable do you think is this CI step? I may have missed something significant, but would it be easier to have a GitHub bot crawling the repos and create the SBOMs for commit hashes, which are currently missing?

**Comment by siberov** [comment]
While the [SBOM-generating action](https://github.com/slsa-framework/github-actions-demo) that this new action depends on is explicitly not production grade, I don't _think_ that this would be unstable, at least as long as we are pinning the version.

Essentially, it builds the project, and hashes the resulting binaries and the source files, storing the result in a [SLSA provenance](https://slsa.dev/provenance/v0.2) file. An example of how the result looks can be seen [here](https://raw.githubusercontent.com/siberov/sbom-files/main/spoon/1665565145-42317ef89a1a1c72987aea5443a4ac0271654d6c.provenance) (warning, large file). Since we want to hash the binaries, just crawling the repo would not be enough.

File: .github/workflows/sbom.yml
Comment (algomaster99) [inline .github/workflows/sbom.yml:11]:
```suggestion
```

Let us make this action trigger only when a commit is pushed to `master`.

**Comment by slarse** [comment]
> To work, this requires that a secret called SBOM_DEPLOY_SSH is added to the Spoon repo. It must contain the private part of an SSH key, where the public key is added as a deploy key to the [chains-project/sbom-files](https://github.com/chains-project/sbom-files) repository.

You should use an environment for this, add that environment to the workflow and also scope that environment only to the master branch. This will limit the risk of having a private SSH key live in the repository workflows.

See the [docs on GitHub Environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment), it's very easy to accomplish and requires only one additional line in the workflow.

**Comment by siberov** [comment]
@slarse , that's a great suggestion, thank you. I was not aware of this feature. If I understand it correctly, it should be addressed by my [latest commit](https://github.com/INRIA/spoon/pull/4955/commits/64419dda4e109773b4283a841ed2a2c970772bd1).

**Comment by slarse** [comment]
> @slarse , that's a great suggestion, thank you. I was not aware of this feature. If I understand it correctly, it should be addressed by my [latest commit](https://github.com/INRIA/spoon/pull/4955/commits/64419dda4e109773b4283a841ed2a2c970772bd1).

Assuming that environment exists and has the desired secret (has to be created through the Settings tab of the repository), then yes that is all you need to do on the workflow side of things :)

**Comment by siberov** [comment]
Hi all,
As @monperrus mentions [here](https://github.com/INRIA/spoon/issues/3879#issuecomment-1280487098), the requisite secrets for this workflow have now been added to the repository.

Does any maintainer have an opinion on where I should go from here to hopefully get this PR merged? If you have any questions at all about the functionality or purpose of this, please don't hesitate to reach out.

**Review by MartinWitt — APPROVED** [review]

**Review by slarse — APPROVED** [review]
@MartinWitt All good by me

**Comment by MartinWitt** [comment]
Okay, before I merge @algomaster99 is the mail `mannu.poski10@gmail.com` as coauthor correct?

**Comment by algomaster99** [comment]
Yes. It's correct.

**Comment by slarse** [comment]
@siberov This CI job is now consistently failing, could you have a look at what's wrong?

**Comment by siberov** [comment]
Yes, I'll look into it.

**Comment by slarse** [comment]
@siberov It's passing again so there's no immediate need to do anything. The error message just made it look like the action wasn't getting the right arguments, but it seems to have just resolved itself.

**Comment by siberov** [comment]
Great to know. I was just about to have a look. Love it when things fix themselves. :)

**Pull request merged**
